Editor: Try a fixed block toolbar

youknowriad opened the pull request:
People are complaining about the block toolbar showing up constantly and disturbing their writing flow, this PR explores the possibility to use a fixed a block toolbar at the top. And its content change depending on the selected block.

<img width="1219" alt="screen shot 2017-08-02 at 10 39 24" src="https://user-images.githubusercontent.com/272444/28867717-e16f1232-776e-11e7-989a-4a2aed6c29e3.png">

Not done yet:

 - The freeform block shows the toolbar differently, we need to tackle it separately
 - Hide this toolbar when the blocks are deselected or when a block doesn't have any control (or maybe show the permalink like shown when focusing the post title)
 - Use this same toolbar for the multiselection toolbar

youknowriad committed:
Editor: Try a fixed block toolbar

jasmussen commented:
Hah, nice, interestingly I just explored something similar with the toolbar in #2151. 

I like this more than I thought I would. I'd like to chew on it for a bit, and perhaps also run it by @afercia. But my gut reaction is positive. CC: @melchoyce @karmatosed

youknowriad commented:
> Hah, nice, interestingly I just explored something similar with the toolbar in #2151.

I was going to say the same in your PR :)

youknowriad commented:
This also has the advantage to reduce the mental-shift when switching from the existing editor (toolbar still at the top)

jasmussen commented:
The primary disadvantage, potentially, could be that some of the context is lost a little bit. This could definitely affect the tie-in to customization, so pinging @melchoyce and @westonruter here. Also @mtias.

youknowriad commented:
> The primary disadvantage, potentially, could be that some of the context is lost a little bit. This could definitely affect the tie-in to customization,

Under the hood, this toolbar can be placed wherever we want to in a different editor. For example, the customization could decide to include it above the blocks, or in a sidebar...

mtias commented:
I don't dislike this (my original concepts when we were starting had a similar concept but at the bottom). But I think it's trying to address a problem that is not of visual design, but of interaction patterns. The UI should not show when you are in "writing" mode, thus not obscuring anything. The problem is how we define "writing" mode.

It sounds like Helen got to a situation where hitting "enter" was showing the UI. That is a bug and we should fix it. Hitting "enter" when you are writing should never show UI.

The other problem is with mousemoves. Perhaps we should disable the functionality that shows the UI when moving the mouse, or make sure that you hover the toolbar area, or add a delay.

I think we should focus on these very evident interaction shortcomings and fix them before making any call regarding the visual design.

StaggerLeee commented:
I dont like it.
Just slight page scroll and User will not know where is his/her ass, and what they are editing.

All editors similar to Gutenberg show toolbar where it is now. Do not overcomplicate it.
But, if it solves for you all hardness of coding for mobile and smaller devices, lack of space, go for it, why not.

afercia commented:
I'd tend to agree with @mtias. Maybe worth considering to better define what "writing mode" is. I'd also consider a shortcut to make the toolbar appear _and receive focus_ on demand, see #552.

The big problem I see with a fixed toolbar is keyboard interaction. Where the toolbar is placed exactly in the source order? How I move focus to the toolbar using only the keyboard?

karmatosed commented:
I have for a while had a lot of issues with the current toolbar because of a simple app I use called PopClip (it has a lot of things that help me interact with content, accessibility things and also just things that make life easier), this is the experience I get:

![2017-08-02 at 14 56](https://user-images.githubusercontent.com/253067/28877763-3c58e044-7795-11e7-9499-a6cfd12ff42d.png)

Now I don't suggest we design for my not so common case, but it has always meant the toolbar was a bit of a bothersome issue for me.

Flipping the thought process, I also prefer design wise having as much as possible context to actions. I can see that being lost the further things get from the content.

One thing I really like that I think covers best of both worlds is a 'shadowing toolbars'. These appear where you need, when you need. A basic level of this is one that follows you around at the top, away from the blocks. A more complex is similar to what @mtias is saying where it appears just when needed.

A little caution here though as to a lot of people saying what a user will or won't do. We are seeing in tests what they are doing, lets maybe focus on that and a solution which solves this.

afercia commented:
> If we don't have enough, lets run some tests. We could also usability test some different options, we have that option.

I'd love to see some user testing sessions include also _at least_: 1 keyboard user, 1 screen reader user, 1 speech recognition software user 🙂

karmatosed commented:
@afercia as would I. Can you help us recruit them?

youknowriad commented:
Worth noting Ghost used this approach but with a toolbar at the bottom.

youknowriad commented:
Another small advantage I'm finding here is the simplicity of the technical solution. Our actual solution is not that simple, it requires us to know if the mouse is moving, if we're typing, if we stopped typing... while this is way simpler, we only need to know the selected block.

Experience shows that often the simplest solutions are the better ones.

mtias commented:
@youknowriad not sure I agree, because I'd still like the UI to not be visible when I'm typing even if it's fixed top/bottom.

afercia commented:
> Can you help us recruit them?

@karmatosed will ask around :)

afercia commented:
> Worth noting Ghost use this approach but with a toolbar at the bottom.

@youknowriad I'd like to move a lot of things at the bottom 🙂  #467

melchoyce commented:
I imagine the inspector will end up being a lot more important than floating toolbars when it comes to customization and layout, with the exception of page/post content (which we might explore making editable via Gutenberg at first anyway). I do find the current implementation a little clunky, so I think @mtias's suggestions are all thing we should try, for sure.

However, I also wouldn't object to us trying out a docked toolbar, maybe more like this with it fixed and centered with the content, either on the top or the bottom:

![image](https://user-images.githubusercontent.com/2846578/28931564-8e521c94-7844-11e7-86be-ad435cf6c3b9.png)

(Though I do worry it could get lost on the bottom)

rianrietveld commented:
User test results https://github.com/WordPress/gutenberg/issues/2259

afercia commented:
> @afercia as would I. Can you help us recruit them?

@karmatosed @rianrietveld is going to ping you about this, she has some volunteers :)

youknowriad commented:
What about this PR? Should we close it?

I really think we should go this road (whether show a fixed toolbar at the top or the bottom, but without centring it IMO) but this is probably not a priority right now and we could revisit later.

jasmussen commented:
I think we should keep it open! I still like it. I don't yet know if it's a perfect fit for us. Even if we need to re-do the PR down the road if we decide to do it (because of rebases), I think it's worth having this open, as it communicates _yes, we are considering this_. Which we are ;)

youknowriad commented:
Another argument for this is that we could decide to always hide the block surroundings (borders, arrows, delete, and cog buttons) unless the block is hovered.

jasmussen commented:
There's also good feedback in this analysis: https://go.tinymce.com/blog/gutenberg-editor-success-lies-timing/ — which would be partially addressed by a docked toolbar.

jasmussen commented:
CC: @annaephox

youknowriad commented:
I pushed this exploration a bit further hiding the surroundings unless we hover a block.

jasmussen commented:
I'm still digging this, even though I also think the current toolbar can work for us. In other words, if popular opinion and tests suggest this is it, then I'm all onboard.

I do think, though, that we'll probably want to always show the toolbar, even if nothing is selected. We'll want to think about what to show there instead — a tip? Seems like we should be able to solve this in a clean way. 

Alternately, we'd want to test whether it was sensible that the toolbar expands like to does — we might even want to add a little animation so it very quickly appears on click (like .1s appear animation). 

Finally, I think we'd want to center the toolbar items in the box. Like provide a 700px maxwidth to a centered container. 

Nice work!

folletto commented:
> I do think, though, that we'll probably want to always show the toolbar, even if nothing is selected. We'll want to think about what to show there instead — a tip? Seems like we should be able to solve this in a clean way.

I've skimmed through the discussion so I might have missed it, but: what if we move the controls inside the existing top bar, and show them only when something is selected? No extra bar, no extra animations, and on mobile it could still pop-out to be a separate bar for space reasons.

Quick mock:

![screen shot 2017-09-29 at 13 03 07](https://user-images.githubusercontent.com/4389/31015166-a3073954-a516-11e7-8ca8-987f56863879.png)

afercia commented:
As already mentioned in a previous comment, this change would significantly worsen the keyboard navigation experience. I'd like everyone here to test with keyboard only (or imagine keyboard interaction with this change) before proposing a fixed toolbar.

> The big problem I see with a fixed toolbar is keyboard interaction. Where the toolbar is placed exactly in the source order? How I move focus to the toolbar using only the keyboard?

Imagine a post with multiple blocks. Should keyboard users then navigate through multiple posts before going back to the toolbar on the top? Or there should be some focus management (which I see very problematic)? How it is supposed to work? Anyone has taken this into consideration? 

The best option is having the controls "in place", as close as possible to the area users are editing, and easy to discover and use. Worth noting the actual interaction with the block toolbars is still problematic for many other reasons, but this is out of the scope of this issue and already tracked on other issues.

youknowriad commented:
> Imagine a post with multiple blocks. Should keyboard users then navigate through multiple posts before going back to the toolbar on the top? Or there should be some focus management (which I see very problematic)? How it is supposed to work? Anyone has taken this into consideration?

I was thinking of something similar to inspector, a button close to the block and a shortcut to jump to the toolbar maybe

folletto commented:
> Anyone has taken this into consideration?

Yes, it's an important consideration. I agree that must be taken in consideration.

I started with the assumption that the current floating toolbar works, as I trusted it was already reviewed by you. If it does, the only difference would be that instead of showing attached visually, would be top aligned. There's no need to mutate DOM and JS behaviour. The only change could be placement.

One of the possible solutions could be CSS only, by having some of the `position: relative` to not constrain the positioning of the toolbar, thus making it placeable in the top bar. Let's find a solution together? :)

afercia commented:
> I was thinking of something similar to the inspector, a button close to the block and a shortcut to jump to the toolbar maybe

I'm afraid it's a too complicated and unintuitive interaction. Also, how to jump back from the toolbar to the block currently being edited? Not a surprise (for sure not for me) we're starting to get feedback and comments like #2801:
> The editor is rather inaccessible. I am blind, so use a screen reader to read my computer. I tested with orca on linux so far, and it... failed pretty miserably. While tabbing around the interface, buttons would seem to be there then not there at random intervals, and some controls I could access (for example the disabled move buttons when you only have one block) suddenly disappeared entirely upon adding another block ...

afercia commented:
> I started with the assumption that the current floating toolbar works

That works as placement in the source order. What doesn't work is the continuous appearing/disappearing of the toolbars, which is highly confusing for users.

> The only change could be [CSS] placement.

Interesting. While for the WCAG the placement in the source order must match the visual order, maybe it's something worth considering, at one condition:

The more I think at navigation through blocks and interaction with them, the more I'm convinced that interaction with blocks should be standardized, see
Keyboard interaction: standardize the blocks behavior #2031

Once that would be done, the following step would be establishing a sort of "Edit mode":
- navigation with keyboard focuses only the blocks
- once users reach the block they want to edit, some command (Enter?) switches the block in Edit move, move focus inside of it, toolbar appears
- when a block is edited, the rest of the interface becomes "inert". the only focusable controls are the edited block and its toolbar (wherever it is placed) and that's the condition mentioned above
- once done editing, a command (Escape? a "Done" button?) exit the block from Edit mode
- navigation through block can continue with the Tab key
- navigation through the toolbar controls still to define, in my opinion it should work like an ARIA toolbar: just one Tab stop and all the controls reachable only with arrow keys

folletto commented:
> The more I think at navigation through blocks and interaction with them, the more I'm convinced that interaction with blocks should be standardized

100% agreed. My perspective is that the idea of blocks is to bring in also that kind of level of standardization. :)

mtias commented:
This is something we have been back and forth in design so much, that I think it might warrant a quick toggle setting so we can test for real and get a better impression of the tradeoffs. I've heard very solid arguments for both proposals, and I think we are exhausting what we can accomplish in discussion form. I also don't feel there's a clear winner yet, so would like to give us the room to test the two.

@youknowriad do you think we could target the toolbar easily to two slots based on a setting?

youknowriad commented:
> @youknowriad do you think we could target the toolbar easily to two slots based on a setting?

I think it's doable yes, with some hacky CSS though. What kind of setting are you thinking of? An URL setting? a toggle somewhere in the UI? I'll probably have to rebuild this from scratch but not a big deal.

youknowriad commented:
Also which one of the "fixed" toolbar design do you think is best? 

- Fixed not centered (could also be at the bottom like ghost)
- Fixed centered but with a 100% border 
- Fixed centered but with a border limited to the content area https://github.com/WordPress/gutenberg/pull/2148#issuecomment-320016104
- Fixed and centered in the header https://github.com/WordPress/gutenberg/pull/2148#issuecomment-333108420

jasmussen commented:
I like these two:

- Fixed centered but with a border limited to the content area https://github.com/WordPress/gutenberg/pull/2148#issuecomment-320016104
- Fixed and centered in the header https://github.com/WordPress/gutenberg/pull/2148#issuecomment-333108420

though I'd prefer a full-width bar instead of the centered bar.

afercia commented:
I can't second this approach of trying new things just considering a visual perspective. As already noted, this change is an accessibility regression and to make it accessible it would require relevant changes that should be included here. They can't be "added later" because there's no guarantee we'll be able to find a good solution. Any relevant UI change should already include accessibility.

jasmussen commented:
My apologies Andrea, I didn't mean to ignore your past statements. I assumed per [your comment here](https://github.com/WordPress/gutenberg/pull/2148#issuecomment-333126084), and per the latest mockup that Yoast provided in the core editor slack which also featured a docked toolbar, that there was a path ahead for this?

afercia commented:
@joen my role in WordPress and my role at Yoast are two separate things :) Coincidentally, those mockups introduce the idea of a sort of "Edit mode", same as in my comment https://github.com/WordPress/gutenberg/pull/2148#issuecomment-333126084

My idea is that in the "Edit mode" the block, the toolbar (and maybe the sidebar) should be the only available and focusable controls and all the rest of the UI should becomes "inert". This could help accessibility as users will be in a sort of "modal" where they have to accomplish a task and the UI makes available only the controls for that task. This part is not mentioned in the mockup, that's a design experimentation. 

My point here is that any experimentation should already include _at least a vague idea_ of how accessibility should be addressed. Instead, what I see is just a visual approach. I'd like to point out again that when there are relevant UI changes, accessibility can't be added later as a bandaid solution, but should be integrated in the design process.

I'm not seeing this happening here, even after months we're collaborating, and that's honestly a bit disheartening.

jasmussen commented:
Thank you for the clarification, very helpful. I'm sorry you feel this way and I completely understand where it is coming from. There's a lot to do here, and sometimes experimentation moves fast — in this case I mistakenly assumed that due to the fact that this was a CSS only change, combined with entering "edit mode" as you described in your comment could be done by simply starting to type or clicking with the mouse, the mockups already could be built in a way that satisfied your suggestions.

However I will try to deeply re-read and think about your comment, and try to make a new mockup as soon as I can. Is there anything else I can do?

afercia commented:
Trying to expand a bit, see screenshot below to get an idea (the opacity layer is just for clarification):

![31015166-a3073954-a516-11e7-8ca8-987f56863879](https://user-images.githubusercontent.com/1682452/31165200-72589064-a8ea-11e7-8bd3-03623e61c164.png)

If all the UI becomes "inert" except the edited block and its toolbar, then the position of the toolbar in the source doesn't matter so much. The only available and focusable controls would be the block itself and its toolbar. The tab order would cycle just through the block and its toolbar.

However, the technical challenge here is making all the rest of the UI not focusable and completely hidden to assistive technologies. I doubt there is a viable solution here and this is the reason why I think this experimentation can't address the underlying accessibility issues.

jasmussen commented:
Very much appreciate your clarification. It sounds a lot like what Google Docs does on mobile:

![screenshot_20171004-100651](https://user-images.githubusercontent.com/1204802/31165686-43505200-a8ec-11e7-9342-16faedd0b260.png)

![screenshot_20171004-100705](https://user-images.githubusercontent.com/1204802/31165689-44937c78-a8ec-11e7-840b-840d1f84aa9b.png)

hedgefield commented:
I like that a lot, and I like the idea of an 'edit' mode. I should clarify on the yoast mockup, including it in the top toolbar there was purely visual. We wanted to try a fixed toolbar but I didn't like leaving so much whitespace so I put it there. But I could definitely see it as the contextual sticky toolbar that @melchoyce suggests above, or as temporarily taking full control of the top bar in 'edit mode' like in @jasmussen's google screenshots. I also wonder if we can't find a middle road by attaching it to the block in a better way.

The thing is that showing the toolbar next to the block all the time is distracting, but if you summon it manually (during some kind of edit mode), it doesn't have to be awkwardly squished between blocks, it can take the space it needs.

![miniblock2](https://user-images.githubusercontent.com/6073772/31181593-06ade30e-a922-11e7-8f25-c965ad08db9f.jpg)

![miniblock1](https://user-images.githubusercontent.com/6073772/31181930-d3a36d84-a922-11e7-8df6-8e89c066473b.jpg)

Attaching it to the block and lifting the whole thing up makes it feel more like a robust whole to me. (For the purpose of this discussion, let's ignore my experimental placement of the icons.)

annaephox commented:
Apologies for jumping in here a little late, hopefully can add in some points that will be useful to this discussion

My reading of the main cause of this issue is that **the toolbar pops up and is distracting**. The proposed solution tabled has been to try fixing the toolbar in some location, mainly to avoid the effects of something popping in and out, and disturbing the writing/reading flow.

A fixed toolbar solution has several complications. Firstly, **we break accessibility**. I won't reiterate the discussion, as it's well articulated above. Secondly, **we break things independent of accessibility** - I ran user tests on something quite similar to this last year, and [we discovered that disconnecting the toolbar from the point of action resulted in 100% user test fails](https://go.tinymce.com/blog/from-the-ux-desk-road-testing-inline-image-editing/).

If we return back to the original cause of the issue, i.e. the toolbar pops up and is distracting. We have user tests that show that [the likely cause of this is the timing with which we are popping up the toolbar](https://go.tinymce.com/blog/gutenberg-editor-success-lies-timing/). Perhaps it is worth exploring tweaking the timing of when the toolbar pops-up as an alternate solution to fixing the toolbar  #(2279)

folletto commented:
Thanks for your feedback Anna.

I think this is worth some clarification, that might have got lost in the discussion. The reason for this suggestion is that:

1. The toolbar covers the content — this is a quite major problem as the editor goal is to provide an interface that allows people to write and edit text. Having something that covers the previous content makes even the simple reading for continuity more difficult. Writing is the core of an editor, so if we can find a way to make it better, we should.
2. The toolbar pops up and distracts — this is related because there are lots of tricks in place to make the floating bar work without distracting (i.e. the work on making it auto-hide while typing), but it's separate because instead is more related to how it's presented. As you noted in the test you link, timing is certainly an issue. But even more, Medium doesn't have that toolbar at all (and has far fewer features, but that's out of scope for this thread).
3. The toolbar keeps moving — this is a minor one, but it plays deep into how perception and cognition work. Having a dedicated area builds up memory over time (so, it's not likely visible on first use), while a moving UI doesn't allow that because every time the person has to re-evaluate where the toolbar is. We minimized this by putting it close to the attention focus of the user, but that depends on how tall is the block.
4. The toolbar might get out of the screen — this is an edge case, but if the toolbar "slides out" then it will have to stick somewhere at the top regardless, like the proposals above, as it would be a problem if it slides out of view. This doesn't apply to small paragraphs, but it's a consideration.

This led us to think alternative ways to present a toolbar that won't cover content and won't distract the user, leading to this thread. The inspiration from this thread is solidly grounded in industry and design standards that evolved over decades: if you recall early apps used floating palettes, but over time more and more apps started having dedicated areas that showed the properties of what's selected. This is a common UI paradigm across desktop apps, and it has also demonstrated its scalability over time. See both consumer and professional apps like Word, Keynote, Photoshop, etc.

Accessibility wise, would be interesting to see for example other companies dealing with the same issue — how did they solve it? I know for example that Word Online has the updating toolbar at the top like the desktop app, and if thats accessible as Microsoft [usually does](https://support.office.com/en-us/article/Learn-how-to-navigate-in-Word-using-accessible-features-a16a93f4-163f-4f48-a246-78c2573d8953#PickTab=Online), maybe we can analyze how they did it :) 

We should try to prototype this one, evaluate how it performs against the current proposal, and if it works, find a way like the one that was [suggested above](https://github.com/WordPress/gutenberg/pull/2148#issuecomment-333121951) to preserve its accessibility.

hedgefield commented:
It got the same feeling of unrest from the points you describe too @folletto.

To point 2: I like the idea of it hiding when you're typing, I think that will help a lot, and will remove the desire for something like an edit mode.

To point 3: I noticed that too. It makes sense to me that it sticks to the top of the window when you scroll down, but sometimes it would stay there if you scroll up again, leaving it in the middle of the block. Fixing that behaviour would already help a lot.

And in general, something about the fact that it hovers slightly over two different blocks at once makes it feel out of place.

folletto commented:
@afercia out of curiosity: how is TinyMCE currently accessible — given it has a static toolbar — and how other apps deal with that (i.e. Word in the online version)?

afercia commented:
@folletto the TinyMCE toolbar is pretty accessible, but it's also immediately before the editing area so it's a bit different case. It implements Alt-F10 to move focus to the toolbar. It also uses `role="toolbar"` which is important for Windows screen readers (e.g. NVDA and JAWS) to correctly pass JS events to the browser. Also, when you make a selection in the edited text, then move to the toolbar, apply some formatting (e.g. bold), focus is moved back to the editor keeping the text selected, which is pretty nice. Selection is preserved also when pressing Escape from the toolbar.

A few considerations:
- since it's an ARIA toolbar (`role="toolbar"`) navigation through the toolbar controls should happen with arrow keys only (no Tab) as specified in the ARIA Authoring Practices expected keyboard interaction: https://www.w3.org/TR/wai-aria-practices/#toolbar
- the toolbar should be labeled
- the TinyMCE demo uses also a menubar (WP doesn't)

Seems the "Toggle Toolbar" button (last one in the first row) breaks keyboard navigation: it is not possible to go through that button with both Tab or Arrow keys. I'm not sure but I think this depends on the WordPress implementation.

You can try it yourself, on the TinyMCE demo: https://www.tinymce.com/
Alt-F9 to focus the menubar
Alt-F10 to focus the formatting toolbar

youknowriad commented:
I'm happy this PR served as a basis to discuss the fixed toolbar alternative. Based on this exploration, a lot of improvements already shipped and let's continue the discussion about the fixed toolbar in #2998
